File: C++/interviewAlgorithms/findlostnum.cpp
//如果要找一串连续数字中，缺失的某个数字，可以直接算这些数的和然后相减，差值就是缺失的那个数
#include<iostream>
#include<vector>
using namespace std;

class FindLostNum{
    public:
        static vector<int> findLostNum(int *array,int length){
            //用于存储2个出现奇数次的整数
            vector<int>result(2);
            //第一次进行整体异或运算
            int xorResult =0;
            for(int i=0;i<length;i++){
                xorResult ^=array[i];
            }
            //如果进行异或运算的结果为0，则说明输入的数组不符合题目要求
            if(xorResult ==0){
                cout<<"输入的数组不符合要求"<<endl;
                return {};
            }
            //确定2个整数的不同位，一次来做分组
            int separator =1;
            while(0 ==(xorResult&separator)){
                separator<<=1;//向左移位，意味着变成原来的2倍（001->010)
            }
            //第二次分组进行异或运算
            for(int i=0;i<length;i++){
                if(0==(array[i]&separator)){
                    result[0] ^=array[i];
                }
                else{
                    result[1] ^=array[i];
                }
            }
            return result;
        }
};

int main(){
    int array[]={4,1,2,2,5,1,4,3};
    //int array[] ={4,4,2,2,6,6,8,8};
    int length =sizeof(array)/sizeof(array[0]);
    vector<int> result=FindLostNum::findLostNum(array,length);
    cout<<result[0]<<","<<result[1]<<endl;
    return 0;
}

File: python/interviewAlgorithms/findLostNum.py
class FindLostNum:
    @staticmethod
    def findLostNum(array):
        # 用于存储两个出现奇数次的整数
        result = [0, 0]
        # 第1次进行异或运算
        xorResult = 0
        for i in range(len(array)):
            xorResult ^= array[i]
        # 如果进行异或运算的结果为0，则说明输入的数组不符合题目要求（两个奇数相同）
        if xorResult == 0:
            return None
        # 确定两个整数的不同位，以此来做分组
        separator = 1
        while 0 == xorResult & separator:
            separator = separator << 1
        # 第2次分组进行异或运算
        for i in range(len(array)):
            if 0 == (array[i] & separator):
                result[0] ^= array[i]
            else:
                result[1] ^= array[i]
        return result

if __name__ == "__main__":
    array = [4, 1, 2, 2, 5, 1, 4, 3]
    result = FindLostNum.findLostNum(array)
    print(str(result[0]) + "," + str(result[1]))
